### PR TITLE
Scc 3096 incorporate time twister

### PR DESCRIPTION
**Pull request by charmingduchess**
In the file dateParser-lambda, I've created:

- timetwist - function to actually send the data to the lambda
- classifyDate - function to classify dates as needing certain transformations 
- preparse - function to apply those transformations in a specific order
- a bunch of functions that are the transformations
- parseDate, the main function 

We're waiting on Alex to do some updates to timetwister for handling seasons, so I've written all the tests regarding that as pending.

### CODE REVIEW

File: lib/dateParser-lambda.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  if (Object.values(dateRange).includes(null)) {
+    preparsedDate = parseOnlyYears(preparsedDate)
+  }
+  return dateRange.map((range) => [range.start, range.end])
+}
+
+const timeTwist = async (preparsedDate) => {
+  const payloadStr = JSON.stringify({
+    path: '/',
+    body: JSON.stringify({ dates: [preparsedDate] })
```
Comment (nonword) [inline lib/dateParser-lambda.js:24]: Note that the date-parser does by design support parsing multiple dates at a time. We should find a way to send all of the dates we want to process in a given batch in single call if possible - to reduce processing time / invocation cost

File: lib/dateParser-lambda.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      case '0X':
+        preparsedDate = squashShortYearBug(preparsedDate)
+        break
+      case 'month-month/month':
+        preparsedDate = monthRangeWithSlash(preparsedDate)
+        break
+    }
+  })
+  return preparsedDate
+}
```
Comment (nonword) [inline lib/dateParser-lambda.js:88]: `classifyDate` applies potentially multiple classifications to a single date. But then `preparse` only applies a single relevant transformation. Was it ever your intent to apply multiple transformations - one for each classification for example? Although that could be powerful, I also worry that could be hard to follow. The structure you have here could support applying multiple transformations based on multiple classifications (e.g. by removing some of those `break` lines), but the execution flow is actually more one-to-one (a single transformation is applied when a classification is found). The trouble is it's kind of hard to determine _which_ transformation is going to be applied because the classifications are built in one place, but the transformation that is applied happens here, and you have to read the switch to figure out which transformation is applied because it's based on the order of cases. So given that the structure is in effect already tree like, I wonder if it would be easier to debug if it was made more explicitly so: rather than build a list of `classifications` (suggesting multiple transformations will be applied), apply the "preparsing" at the moment of classification.

**Comment by charmingduchess** [comment]
Current success rates are reported as 80% for volumes and 63% for dates. I think it is actually higher than that. I found at least one example of a date failing because of differences in specificity between target and parsed (the target was 2002-06 and the parsed was 2002-06-21). I think as far as volume and date parsing goes, we are in great shape.

File: lib/volume-parser.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+exports.parseVolume = (fieldTagV) => {
+  const regExp = [/(?:^|\s|-)(?:volume|vol\.?|v\.?)\s?(\d*)[-|/]?(\d*)?/gi, /(?:^|\s|-)(?:no\.?)\s?(\d*)[-|/]?(\d*)?/gi, /(?:^|\s|-)(?:bd\.?)\s?(\d*)[-|/]?(\d*)?/gi, /(?:^|\s|-)(?:reel|r\.?)\s?(\d*)[-|/]?(\d*)?/gi, /(?:^|\s|-)(?:reel|r\.?)\s?(\d*)[-|/]?(\d*)?/gi]
```
Comment (nonword) [inline lib/volume-parser.js:2]: Can you format this with line-breaks and inline comments about generally what each is supposed to match and or an example?

File: test/volume-parse-test.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  it('vol 93, no. 3 (autumn 2013)', () => {
+    expect(parseVolume('vol 93, no. 3 (autumn 2013)')).to.deep.equal([93, 93])
+  })
+  it('v. 6-7 no. 2, 5-v. 8 no. 1 (Oct. 1961-Sept./Oct. 1962, May-June/July 1963)', () => {
+    expect(parseVolume('v. 6-7 no. 2, 5-v. 8 no. 1 (Oct. 1961-Sept./Oct. 1962, May-June/July 1963)')).to.deep.equal([[6, 7], [8, 8]])
+  })
+  it('v. 6-7 no. 2, 5-v. 8-10 no. 1 (Oct. 1961-Sept./Oct. 1962, May-June/July 1963)', () => {
+    expect(parseVolume('v. 6-7 no. 2, 5-v. 8-10 no. 1 (Oct. 1961-Sept./Oct. 1962, May-June/July 1963)')).to.deep.equal([[6, 7], [8, 10]])
+  })
+})
```
Comment (nonword) [inline test/volume-parse-test.js:24]: Almost all of these will always fail in travis because the environment in travis isn't configured to connect to the dateparser lambda. So can we tag these as local only and exclude them from the standard test run? I think we can do that in mocha by:
 1. Add, for example, "@local-only" to the end of the `it` description
 2. Add ` --grep '@local-only' --invert` to the `mocha` call in the registered test script in `package.json`

**Comment by nonword** [comment]
A few misses that I think we can correct now:
 * `267. Parsing: "v. 463 [1982]"`: Can we fall back as a last resort on matching a single 4 digit year?
 * `263. Parsing: "1895-1896/1897"` This one actually parses correctly, but the target in the spreadsheet is wrong, so we should update it there and update the targets csv
 * `258. Parsing: "v. 24 1883"`: Here again, when all else fails, can we attempt to match a single 4-digit year anywhere in the string? In fact, it looks like there are several cases where a single 4-digit year isn't being found, so correcting that should bump the score by a few points
 * `240. Parsing: "1969/70-1973"`: This one is wrong in the CSV and should be updated via the spreadsheet
 * `219. Parsing: "v. 5-7, no. 37-60 (1963 inc)"`: Also wrong in the csv
 * `189. Parsing: "v. 13-16 1982"`: Also wrong in the csv
 * `161. Parsing: "jaarg. 24 (Jan.-June 1967)"`: Can we support jaarg. as a fallback after Vol?
 * `145. Parsing: "1984-1988/90"`: This is evaluating to `[["0","1988"]]`
 * `140. Parsing: "v. 9 (1938-1939)"` Also wrong in CSV

In general, please be skeptical of failures as the target data may need to be corrected
 
Also, I see a few cases like `Input parsed as:  [["1912","1927"],[null,null]]`. I believe this is happening due to commas. Maybe a post-parsing step to strip arrays of nulls? That should bump the matches percentage.

As discussed, also fine with putting this work into an "improvements" ticket for later.
